src/librustc_mir/transform/simplify_cfg.rs

```diff
@@ -160,8 +160,14 @@ impl<'a, 'tcx: 'a> CfgSimplifier<'a, 'tcx> {
         debug!("collapsing goto chain from {:?} to {:?}", *start, target);
 
         *changed |= *start != target;
-        self.pred_count[target] += 1;
-        self.pred_count[*start] -= 1;
+
+        if self.pred_count[*start] == 1 {
+            self.pred_count[*start] = 0;
```

Comment:
Took me a while to get that this is because `*start` becomes dead here and thus shouldn't count.
There should be a comment making this (or whatever the correct explanation) clear to future readers.